### DISCUSSION TITLE
대표님께서는 Ros2 Foxy와 mcu 사이에 통신할 때 어떤 라이브러리를 사용하시는지 궁금합니다

### DISCUSSION THREAD

**Discussion by Jeoungji** [Road-Balance/gcamp_ros2_basic#72] [Q&A]
안녕하세요
ubuntu 20.04 환경에 대표님께서 강의 때 올려주신 Ros2 Foxy 환경에서 로봇 제작을 진행하고 있습니다.
개발 과정 중 ros와 mcu 사이에서 통신이 문제가 되어 혹시 답변을 얻을 수 있을까 하여 질문남깁니다.

저희 로봇은 ubuntu20.04 환경에서 ros2 foxy 버전을 사용하고 있습니다.
컴퓨터에 센서과 통신에 사용되는 esp32 4개, 모터 제어를 위한 아두이노 듀에 총 5개의 mcu를 연결할 생각입니다.
복잡한 통신은 아니며 단순히 센서 값과 모터 값만 보내고 받을 예정입니다.

esp32를 ros와 통신하기 위하여 micro_ros 라이브러리를 사용해보려고 하였지만 rmw_implementation 빌드 오류 등 2개의 오류가 생겨 패키지 빌드가 되지 않아 사용을 못하고 있으며,
pyserial을 통해 serial통신을 해보려고 하였지만 serial라이브러리가 import되지 않는 문제가 있었습니다.
나머지 하나인 ros2arduino는 ros2 dashing 버전을 지원하는 것 같아 시도해보지는 않았습니다.

혹시 대표님께서는 통신을 위해 어떤 라이브러리를 사용하고 계신지 궁금하여 질문남깁니다.

**Comment by kimsooyoung** [comment]
저는 micro ros를 사용하고 있습니다. rmw_implementation 오류가 무엇인지 정확하게 알 수는 없지만, micro-ros의 개발자분께 지금 처럼 github 질문을 남기면 답변을 잘 주십니다. 저 또한 구현 시 질문을 남겼는데 하루 정도 걸려 바로 답변을 얻었던 기억이 있습니다.

혹은 C++ serial 라이브러리를 사용하는 방법도 있습니다. 하지만 이렇게 되면 기존 ROS 시스템과 혼용할 시 구현이 복잡해지는 단점이 있어요.

정리하자면
* 방법 1 : micro-ros 
장점 : 많은 예시들이 있고 비교적 직접 개발해야 할 것이 적음
단점 : 초기 세팅 시 많은 시행착오 필요, 사용할 수 있는 기기에 제한이 있음 (esp32는 아마 될겁니다.)

* 방법 2 : 별도  serial 라이브러리 사용 ([예시 링크](https://github.com/wjwwood/serial))
장점 : 기기 제한 없이 시리얼 통신만 되면 사용 가능
단점 : 직접 구현해야 하는 부분이 다소 있으며, 구현에 따라 ROS의 동작과 충돌이 일어날 수 있음

화이팅입니다!!

**Reply by Jeoungji** [reply]
감사합니다!! 질문 남겨보도록 하겠습니다!
